File: 2021/qualifications/p3/q3.ml
(* X = ΣelemL1 * Σelem2
 *  len L1 + len L2 is max *)

(** idée pour accélerer (ne réduit pas la complexité)
 ** - enlever le "0"
 ** - une fois qu'on a une première solution, ne faire que des tests avec longeur des listes >
 ** - etudier le signe / primalité de la clé *)
(* Copy String.split_on_char from OCaml 4.04 *)
module String = struct
  include String

  let split_on_char sep s =
    let r = ref [] in
    let j = ref (String.length s) in
    for i = String.length s - 1 downto 0 do
      if String.unsafe_get s i = sep then begin
        r := String.sub s (i + 1) (!j - i - 1) :: !r;
        j := i
      end
    done;
    String.sub s 0 !j :: !r
end

(**
   @param x le nombre magique
   @param n la longueur du code la Matriks
   @param l le code de la Matriks
*)

let rec print_int_list (l:int list) =
  match l with
  | [] -> () (*print_string "]"*)
  | t::q -> print_int t; if not (q = []) then print_string " "; print_int_list q

let rec blited_list (l:'a list) (pos1:int) (pos2:int) =
  (** revoie la liste des élements contenus dans la liste l entre les indices pos1 et pos2 inclus *)
  match l with
  | [] -> []
  | t::q ->
    if pos1 > 0 then
      blited_list q (pos1-1) pos2
    else
    if pos2 > 0 then
      t::blited_list q pos1 (pos2-1)
    else
      []

let rec sum_elt (l:int list) =
  (** revoie la somme des élements contenus dans la liste l*)
  match l with
  | [] -> 0
  | t::q -> t+sum_elt q

let test_cle l x i1 i2 len1 len2 =
  (***)
  sum_elt (blited_list l i1 (len1-i1)) * sum_elt (blited_list l i2 (len2-i2)) = x

let resoudre x n l =
  (** TODO Les deux clés (chacune sur une ligne) ou le message "IMPOSSIBLE". *)
  (* On essaie une méthode brute ... *)

  let best_i1 = ref 0
  and best_i2 = ref 0
  and best_len1 = ref 0
  and best_len2 = ref 0 in
  for i1 = 0 to n do
    for i2 = 0 to n do
      for len1 = i1+1 to n do
        for len2 = i2+1 to n do
          (* if test_cle l x i1 i2 len1 len2 then *)
          begin
            print_string "[";
            print_int_list (blited_list l i1 (len1- i1));
            print_string " ; ";
              print_int_list (blited_list l i2 (len2- i2));
            print_string "]";
            print_newline()
            end

          (* if test_cle l x i1 i2 len1 len2 then
           *   if len1+len2 > !best_len1 + !best_len2 then
           *     begin
           *     best_len1 := len1;
           *     best_len2 := len2;
           *     best_i1 := i1;
           *     best_i2 := i2;
           *     end *)
        done;
      done;
    done;
  done;
  if !best_i1 = 0 && !best_i2 = 0 && !best_len1 = 0 && !best_len2 = 0 then
    print_string "IMPOSSIBLE"
  else
    let l1 = blited_list l !best_i1 (!best_len1- !best_i1)
    and l2 = blited_list l !best_i2 (!best_len2- !best_i2) in
    (** l1 > l2*)
    if List.length l1 > List.length l2 then
      begin
        print_int_list l1;
        print_newline();
        print_int_list l2;
      end
    (** l2 > l1*)
    else if List.length l2 > List.length l1 then
      begin
        print_int_list l2;
        print_newline();
        print_int_list l1;
      end
    else
      (** l1 = l2*)
    (** sl1 > sl2*)
    if sum_elt l1 > sum_elt l2 then
      begin
        print_int_list l1;
        print_newline();
        print_int_list l2;
      end
    else
      begin
        print_int_list l2;
        print_newline();
        print_int_list l1;
      end


let () =
  let x = read_int () in
  let n = read_int () in
  let l = read_line () |> fun x -> if x = "" then [] else String.split_on_char ' ' x |> List.rev_map int_of_string |> List.rev in
  resoudre x n l

File: 2021/qualifications/p3/q3.py
from math import sqrt
def get_cumulative_tab (l, n):
    tab = n*[0]
    s = 0
    for i in range(n):
        s += l[i]
        tab[i] = s
    return tab

def tab_somme_elts(tab, cum_tab, n):
    '''tableau dont les élement sont les sommes des sous-listes'''
    tab_sum = []
def get_sommes(cum_tab, x):
    '''Donne les deux sommes solution ou renvoit -1'''
    # -> on coupe en deux la liste cum_tab donnant l1 et l2, une avec les élements > sqrt(x), l'autre avec le reste
    # on cherche d'abord l'indice auquel on veut séparer les listes
    i_sep = len(cum_tab)             # cas où tous les élements de la liste sont inf à sqrt(x), la 2e liste et vide et on ne poura pas obtenir x
    for i in range(len(cum_tab)):
        if cum_tab[i] > int(sqrt(x)) and i<i_sep:
            i_sep = i

    l1 = cum_tab[:i_sep]
    l2 = cum_tab[i_sep:]
    print(cum_tab)
    print(i_sep)
    print(l1)
    print(l2)
    for s2 in l2:
        for s1 in l1:
            print(f"s1:{s1} s2:{s2} produit:{s1*s2}")
            if s1*s2 == x:
                return [s1,s2] # s1 < s2
    return -1

def sousliste_of_somme(l, cum_tab, s, n):
    best_list = []
    best_len = 0
    for pos1 in range(n):
        for pos2 in range(n):
            if cum_tab[pos2] - cum_tab[pos1] -1 == s:
                if pos2 - pos1 > best_len:
                    best_list = l[pos1:pos2+1]
                    best_len = pos2-pos1
    return best_list

def resoudre(x, n, l):
    """
    :param x: le nombre magique
    :type x: int
    :param n: la longueur du code la Matriks
    :type n: int
    :param l: le code de la Matriks
    :type l: list[int]
    """
    # TODO Les deux clés (chacune sur une ligne) ou le message "IMPOSSIBLE".
    cum_tab = get_cumulative_tab(l, n)
    print(cum_tab)
    best_pos = [0,0]
    sommes = get_sommes(cum_tab, x)
    print(sommes)
    s1 = sommes[0]
    s2 = sommes[1]
    l1 = sousliste_of_somme(l, cum_tab, s1, n)
    l2 = sousliste_of_somme(l, cum_tab, s2, n)
    print(l1)
    print(l2)

if __name__ == '__main__':
    x = int(input())
    n = int(input())
    l = list(map(int, input().split()))
    resoudre(x, n, l)
